### PR TITLE
Support casting from binary to dictionary of binary

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Part of #1179.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-cast/src/cast.rs [hunk 1]
```diff
@@ -3348,6 +3350,30 @@ where
     Ok(Arc::new(b.finish()))
 }
 
+// Packs the data as a BinaryDictionaryArray, if possible, with the
+// key types of K
+fn pack_binary_to_dictionary<K>(
+    array: &ArrayRef,
+    cast_options: &CastOptions,
+) -> Result<ArrayRef, ArrowError>
+where
```
Comment (viirya) [inline arrow-cast/src/cast.rs:3359]: Perhaps we can have a generic one for both string and binary, I may try it in a follow up work.

File: arrow-array/src/builder/generic_bytes_dictionary_builder.rs [hunk 2]
```diff
@@ -268,7 +268,7 @@ where
         let keys = self.keys_builder.finish();
 
         let data_type =
-            DataType::Dictionary(Box::new(K::DATA_TYPE), Box::new(DataType::Utf8));
+            DataType::Dictionary(Box::new(K::DATA_TYPE), Box::new(T::DATA_TYPE));
```
Comment (viirya) [inline arrow-array/src/builder/generic_bytes_dictionary_builder.rs:271]: Found a bug.

Comment (tustvold) [reply arrow-array/src/builder/generic_bytes_dictionary_builder.rs:271]: Eep

File: arrow-cast/src/cast.rs [hunk 3]
```diff
@@ -3285,6 +3285,8 @@ fn cast_to_dictionary<K: ArrowDictionaryKeyType>(
         ),
         Utf8 => pack_string_to_dictionary::<K>(array, cast_options),
         LargeUtf8 => pack_string_to_dictionary::<K>(array, cast_options),
+        Binary => pack_binary_to_dictionary::<K>(array, cast_options),
+        LargeBinary => pack_binary_to_dictionary::<K>(array, cast_options),
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:3289]: The method assumes BinaryArray, so I'm not sure how this works?

File: arrow/tests/array_cast.rs [hunk 4]
```diff
@@ -411,7 +411,9 @@ fn get_all_types() -> Vec<DataType> {
         ),
         Dictionary(Box::new(DataType::Int8), Box::new(DataType::Int32)),
         Dictionary(Box::new(DataType::Int16), Box::new(DataType::Utf8)),
+        Dictionary(Box::new(DataType::Int16), Box::new(DataType::Binary)),
         Dictionary(Box::new(DataType::UInt32), Box::new(DataType::Utf8)),
+        Dictionary(Box::new(DataType::UInt32), Box::new(DataType::Binary)),
```
Comment (tustvold) [inline arrow/tests/array_cast.rs:416]: What's the rationale for this choice of dictionary keys

File: arrow-cast/src/cast.rs (hunk 3 above)
Comment (viirya) [reply arrow-cast/src/cast.rs:3289]: Like `pack_string_to_dictionary`, they both perform casting first to get `StringArray`/`BinaryArray` which is taken to build dictionary array. I think we can simplify this by removing first casting. I will do it in a follow up pr.

File: arrow/tests/array_cast.rs (hunk 4 above)
Comment (viirya) [reply arrow/tests/array_cast.rs:416]: I just picked it following other existing types. Maybe as a improvement, I can refactor this in a separate pr to iterate all possible dictionary key types.

File: arrow-cast/src/cast.rs (hunk 3 above)
Comment (tustvold) [reply arrow-cast/src/cast.rs:3289]: Yeah this is also problematic as it bat truncate offsets, would be good to fix

**Review by tustvold — APPROVED** [review]
Sorry for the delay

**Pull request merged**